### DISCUSSION TITLE
[Suggestion] Add 2FA (Two-factor authentication) to Accounts

### DISCUSSION THREAD

**Discussion by Soulsderfuchs** [RetroAchievements/RAWeb#809] [Ideas]
Currently the site uses basic logging functionality ,making breaching of an account easily done if a third party gets hold of the password to said account and in a worst case scenario being capable of requesting deletion of the account or reseting progress on sets.

Adding optional security options such as 2FA would make these accounts more secure for the users.

**Comment by televandalist** [comment]
Here's hoping...

**Comment by Sanaki** [comment]
I'm good with it. It -might- be doable in V1, but someone more familiar with the codebase would have to weigh in. My only fundamental concern is that it needs to be via TOTP (google authenticator and authy being the most common providers), because email-based is the only other viable option, and that's both less secure and less reliable, with some of our users having email issues already on V1.

Topics I see that'd need to be addressed:
* Just implementing it. Straightforward.
* Emulator support. For this one, a few options I see, most with issues...
  * Add a TOTP field for login (not great, since RetroArch login is asynchronous with entry)
  * Allow appending TOTP to the password (not great, see above)
  * Allow a timed disabling of TOTP from the website to fetch a token (over-involved for users, counter to TOTP security)
  * Just ignore TOTP from game clients (most viable, least secure)
  * Allow generating a one-use client password after site login which can be used to log in a single game client (probably the best option available, but may push it out of V1 viability)
  * Update RetroArch login to attempt login through settings, rather than on hash validation. Add a TOTP field in the process.
* Password/account recovery policy becomes something we need to seriously contemplate, since users will inevitably lock themselves out (primarily through user error).

**Reply by MichaelRPowell** [reply]
>   * Allow generating a one-use client password after site login which can be used to log in a single game client (probably the best option available, but may push it out of V1 viability)

I think creating some type of key that can be used in RetroArch or standalone emulators to allow achievements to be pushed to the service would be great. The key would be limited to only adding achievements or updating leaderboard scores.

I personally would be okay with a single key being allowed on multiple devices and would be easiest for me. Though agreeably that would not be the most secure. I do not have any issue going in and removing faulty achievements earned if someone were to abuse the key or if it somehow was made public.

I think the bigger concern most people have is having their accounts wiped. Two-factor authentication would be nice to have to help prevent that and other malicious changes.

**Comment by FOSSilizedDaemon** [comment]
I would be extremely happy if this got done!

**Comment by luchaos** [comment]
Preparations for TOTP for site authentication have been made and will be worked on further as soon as the V2 base branch was released.

It will be compatible with all common providers (using QR codes), provide recovery paper keys and session management in a separate section in the then overhauled settings area.

Related to that, an OAuth server instance will allow for "Log in via RetroAchievements.org" for third parties and will also include a replacement for Web API tokens using OAuth personal tokens. 

Emulator authentication for the Connect/Integration RPC API is a different beast as it would have to play nicely with both the 2FA and the OAuth layer within the clients. Will certainly not be included in the first pass of this feature; needs a lot more planning and thought as @Sanaki mentioned.

**Reply by FOSSilizedDaemon** [reply]
Awesome!
